WIP: MachineConfigPool Surge and NodeDrainTimeout Support (#1616)
openshift/enhancements: enhancements/update/standalone-worker-surge.md

---
title: machine-config-pool-update-surge-and-nodedraintimeout
authors:
  - jupierce
reviewers: 
  - TBD
approvers: 
  - TBD
api-approvers: 
  - TBD
creation-date: 2024-04-29
last-updated: 2024-04-29
tracking-link:
  - TBD
see-also:
  - https://github.com/openshift/enhancements/pull/1571
---

# MachineConfigPool Update Surge and NodeDrainTimeout

## Summary

> Add `MaxSurge` and `NodeDrainTimeout` semantics to `MachineConfigPool` to improve the predictability 

Author: You can find these values in CAPI MachineDeployment today: https://github.com/Nordix/cluster-api/blob/9a2d8cdc5ad681ba407e47106cb159bfa708763c/config/crd/bases/cluster.x-k8s.io_machinedeployments.yaml . 
I'm assuming we will not be suggesting direct manipulation of `MachineDeployment` once we are on CAPI; i.e. we will still be using `MachineConfigPool` as the abstraction / customer touch point for this type of configuration.

> Add `MaxSurge` and `NodeDrainTimeout` semantics to `MachineConfigPool` to improve the predictability 

Reviewer: > I'm assuming we will not be suggesting direct manipulation of MachineDeployment once we are on CAPI; i.e. we will still be using MachineConfigPool as the abstraction / customer touch point for this type of configuration.

I disagree. We are expecting users to move towards MachineDeployment instead of MachineSets to provide greater abstractions and upgrade features such as RollingUpdates. MachineConfigPool is really about the software that is configured on a group of nodes, the MachineDeployment IMO is the one that specifies how that gets onto the nodes. Whether that is in-place or a rolling update

node capacity available to schedule new pods to replace the pods being drained. This can result in a stalled drain -- one that 
does not terminate until there is external intervention.

Stalled drains create a frustrating experience for operations teams as they require analysis and intervention. They can also
make it impossible to predict when an update will complete -- complicating work schedules and communications. There are a number of reasons 
drains can stall, but simple lack of spare worker node capacity is a common one. The reason is that running an over-provisioned cluster
(i.e. with more worker nodes than actually required for the steady state workloads) is not always cost-effective. Running
an extra node 24x7 simply to ensure capacity for the short period while another node is being drained is inefficient.

One cost-effective approach to ensure capacity is called "surging". With a surge strategy, during an update, the platform
is permitted to bring additional worker nodes online, to accommodate workloads being drained from existing nodes. After an 

> update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

Reviewer 2: In OSD/ROSA we scale all machinesets by creating a new machineset with replicas=1 that matches the original machineset in all but name (we append some string).  In this proposal I'd expect as a machineset is being updated the same happens, blindly scaling everything by the surge amount.

Agree on the disruption!  If it were possible to reduce that it would be great.  A challenge we have had for managed openshift to date is machineset is not immuable while upgrades happen, hence the new machineset creation to isolate from any changes that might come in while an upgrade is in progress.  This means we can't get away from this disruption.  If surge were a first class citizen and this tweak to simply destroy non-updated machines to get back to the non-surge replicas it would be very nice.

> update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

Author: I am adding implementation details after further discussion with lifecycle that mirror what OSD has done. This should address the machineset shape question directly.
For disruption, I don't think any update mechanism tries to prevent double disruption (of non-HA workloads). Unless the kube-scheduler can add weight to "newest nodes", there is every chance a drained workload will find a seat on a old node that is, itself, soon to be drained..

> update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

Reviewer: > In OSD/ROSA we scale all machinesets by creating a new machineset with replicas=1 that matches the original machineset in all but name

You just mean during upgrades right? Not during normal operation?

> If surge were a first class citizen and this tweak to simply destroy non-updated machines to get back to the non-surge replicas it would be very nice.

I believe that is the behaviour of the Cluster API MachineDeployment already, do you know if that is the behaviour on HCP?

> For disruption, I don't think any update mechanism tries to prevent double disruption (of non-HA workloads). Unless the kube-scheduler can add weight to "newest nodes", there is every chance a drained workload will find a seat on a old node that is, itself, soon to be drained..

Yeah true, but I think the current suggestion of scaling up and then down again makes it more likely/perhaps opens us up to triple?

> update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

Reviewer 2: Yes, only in upgrade.

In HCP we're using default "Replace" behavior.  I hope it's not updating a node just to delete one.  Asking to confirm, will follow up here.

I haven't heard a requirement to optimize for least disruption of workloads.  That feels like a larger thing to tackle and is not a new problem brought on by this enhancement.  Scheduling to a yet-to-be-updated node is going to happen today 100% of the time for the first node update with additional chances on subsequent nodes.

> update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

Reviewer: > I haven't heard a requirement to optimize for least disruption of workloads. That feels like a larger thing to tackle and is not a new problem brought on by this enhancement. Scheduling to a yet-to-be-updated node is going to happen today 100% of the time for the first node update with additional chances on subsequent nodes.

Yes agreed, but, if we know that this is a problem and we are making choices we know make this worse, do we really want to make those choices? We don't need to fix it now, but I'd have thought we would not want to add something we _know_ makes the situation worse

> update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

Reviewer 2: > I'd have thought we would not want to add something we know makes the situation worse

Would it be a considered a regression for this behavior to change and for restarts of workloads to happen more often?  If not, who makes the call to make that a concern we're willing to consider for regression?

> update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

Reviewer: I think that depends on whether this is something we test. I know that we do test disruption of workloads during upgrades today. For example, there are load balancer tests with example workloads that test how much downtime a customer application would face as we go through an upgrade.
Currently, because that works with PDBs and the LB health checks for the various clouds, I think that this wouldn't necessarily fall foul of those.

But I'd be interested to see what the TRT folks think, since they are the ones who look after our testing and pick up regressions like this.

HyperShfit Hosted Control Planes (HCP) already support the surge concept. HyperShift `NodePools`
expose `maxUnavailable` and `maxSurge` as configurable options during updates: https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.RollingUpdate . 

> Unfortunately, standalone OpenShift, which uses `MachineConfigPools`, does not. To workaround this

Reviewer: I think it would probably be preferable all round to not duplicate the features here since we want MachineDeployment in the standalone product in the future. Perhaps instead of spending cycles implementing this here, the time could be spent helping to bring MachineDeployment in and solving the issues within OCP that this would cause (like the MCS being able to serve a particular new ignition version of the rendered configuration)

> Unfortunately, standalone OpenShift, which uses `MachineConfigPools`, does not. To workaround this

Reviewer 3: Hypershift does indeed replace all workers. I think one thing we can look to do in standalone is allow a hybrid update model, where you can configure maxSurge and maxUnavailable on top.

What I mean by that is, let's say you have 6 worker nodes, and you want the end result to be 6 worker nodes, and you don't really care how that's achieved, so long as:
1. you never go above 9 workers (maxSurge of 3), and
2. 4 workers are always available for scheduling (maxUnavailable of 4)

In that case something in the cluster can configure it such that:
1. some nodes are updated inplace, in parallel
2. some nodes are surge'd in to the cluster on the newest configuration directly
3. some nodes are simply removed

In that case this something needs to preserve the maxSurge and maxUnavailable we set above. Could that be done via a hybrid of CAPI and MCO? Or should there be a central management point for this?

> limitation for managed services customers, Service Delivery developed a custom Machine Update Operator (MUO)

Reviewer 2: ```suggestion
limitation for managed services customers, Service Delivery developed a custom Managed Ugrade Operator (MUO)

```

which can surge a standalone cluster during an update (see [reserved capacity feature](https://github.com/openshift/managed-upgrade-operator/blob/a56079fda6ab4088f350b05ed007896a4cabcd97/docs/faq.md)).

### Preventing Other Stalled Drains

As previously mentioned, there are other reasons that drains can stall. For example, `PodDisruptionBudgets` can

> be configured in such a way as to prevent pods from draining even if there is sufficient capacity for them

Reviewer 2: When self managed it's fine for the admin to handle this on a case by case basis.  For RH managed openshift it cannot scale, we have to automate this.  And we don't know customer workloads, actually cannot even see them.  Best (only) thing we can do is ensure the upgrade progresses after a reasonable and documented waiting period.

> be configured in such a way as to prevent pods from draining even if there is sufficient capacity for them

Reviewer: I know I'm about to cross into a whole other territory here, but bear with me.

IIUC, the other enhancement currently is talking about separating control plane and worker upgrades in a way such that our end users have greater control over when they are disrupted right? I'm wondering if the "policy" for force upgrading should be a part of that configuration?

Users can't run workloads on the control plane right so I don't think this is a control plane issue.

Once the other EP is merged, how much do we care about workload nodes being upgraded, how much drift can there be without it causing headaches?

> be configured in such a way as to prevent pods from draining even if there is sufficient capacity for them

Reviewer 2: Customers can set the timeout for forcing drain today at a cluster level.  For drift we care about infra nodes, which managed openshift does deploy.  Those should not fall behind.  Should be a max allowed skew on control plane upgrade.  I'll check the other EP for both of these..

> be configured in such a way as to prevent pods from draining even if there is sufficient capacity for them

Author: I separated the enhancements because I think they have value in isolation, but it is most impactful for SD if we execute on both (i.e. they can dramatically reduce the MUO). I would still suggest separate configuration stanzas:
1. It allows this enhancement config stanza to mirror HCP and MachineDeployment.
2. It is valid and simplifying to separate (a) **when** an update should be performed (https://github.com/openshift/enhancements/pull/1571) and (b) **how** an update should be performed (this enhancement). 
3. No one wants the maintenance schedule enhancement to be even longer :) .

> be configured in such a way as to prevent pods from draining even if there is sufficient capacity for them

Author: > Once the other EP is merged, how much do we care about workload nodes being upgraded.
Much less. I expect most customers will update worker nodes less often. 

> how much drift can there be without it causing headaches?
N-2 today and N-3 tomorrow when we adopt upstream kube's posture.

> be configured in such a way as to prevent pods from draining even if there is sufficient capacity for them

Reviewer: I agree it makes sense to separate the two, but I'm wondering if this becomes much simpler if we assume that the maintenance schedules already exist (ie doing that one first and implementing this after)

> N-2 today and N-3 tomorrow when we adopt upstream kube's posture.

So realistically a maximum drain timeout could be set to a year and not cause an issue, interesting 🤔

to be rescheduled on other nodes. A powerful (though blunt) tool to prevent drain stalls is to limit the amount of time
a drain operation is permitted to run before forcibly terminating pods and allowing an update to proceed. 
`NodeDrainTimeout`, in HCP's `NodePools` allows users to configure this timeout.

> The Managed Update Operator also supports this feature with [`PDBForceDrainTimeout`](https://github.com/openshift/managed-upgrade-operator/blob/master/docs/faq.md).

Reviewer 2: ```suggestion
The Managed Upgrade Operator also supports this feature with [`PDBForceDrainTimeout`](https://github.com/openshift/managed-upgrade-operator/blob/master/docs/faq.md).

```

Author: lol - I'll get it right one of these times.

This enhancement includes adding `NodeDrainTimeout` to `MachineConfigPools` to provide this feature in standlone 
cluster environments.

### User Stories

Implementing surge and node drain timeout support in `MachineConfigPools` can simplify cluster management for self-managed 
standalone clusters as well as managed clusters (i.e. Service Delivery can remove this customized behavior from the MUO and use
more of the core platform).

* As an Operations team managing one or more standalone clusters, I want to
  help ensure smooth updates by surging my worker node count without constantly 
  having my cluster over-provisioned.
* As an Operations team managing one or more standalone clusters, I want to 
  ensure my cluster update makes steady progress by limiting the amount of time a node drain can

>   consume.

Reviewer 2: Is the intent to force drain a node only in the context of an upgrade?  What of other operations such as reaping a node in MHC, would that trigger a forced drain?  MUO today and SD's forced drain only apply in upgrade.

Author: Great catch. Adding clarifications that this is only the context of a automated upgrade operations. Other callers to cordon/drain must pass their own parameters.

* As an engineer in the Service Delivery organization, I want to use core platform

>   features instead of developing, evolving, and testing the Managed Update Operator.

Reviewer 2: ```suggestion
  features instead of developing, evolving, and testing the Managed Upgrade Operator.

```

* As an Operations team managing standalone and HCP based OpenShift clusters, I

>   want a consistent update experience leveraging a surge strategy and/or 

Reviewer 2: HCP does replace nodes by default, I believe you can opt for in-place updates also.  For managed openshift we use replace.  I'm +1 to getting the option for replace for classic if that's the direction OCP wants to go.  We do have some customers in HCP that require in-place upgrades if I remember correctly, so it needs to be configurable.

>   want a consistent update experience leveraging a surge strategy and/or 

Reviewer: As far as I'm concerned, yes, it is something we want to bring in. I have this on the discussion topics already for the CAPI F2F end of July where I'm hoping to get more consensus on how we are going to handle upgrades (CVO/MCO) in a more CAPI driven OpenShift land.

I imagine that yes, in the future, both HCP and OCP will support both, and it would be preferable I think to leverage the same building blocks to do so, rather than inventing something new

  node drain timeouts regardless of the cluster profile.

### Goals

- Implement an update configuration, including `MaxSurge`, similar to HCP's `NodePool`
  in standalone OpenShift's `MachineConfigPool`. 
- Implement `NodeDrainTimeout`, similar to HCP's `NodePool`, in standalone OpenShift's `MachineConfigPool`.
- Provide a consistent update controls between standalone and HCP cluster profiles. 
- Allow Service Delivery to deprecate their MUO reserved capacity & `PDBForceDrainTimeout` features and use more of the core platform.

### Non-Goals

- Address all causes of problematic updates.

> - Prevent workload disruption when `NodeDrainTImeout` is utilized.

Reviewer 2: ```suggestion
- Prevent workload disruption when `NodeDrainTimeout` is utilized.

```

### API Extensions

#### API Overview
The Standalone `MachineConfigPool` custom resource is updated to include new update strategies (one of which 
supports`MaxSurge`) and `NodeDrainTimeout` semantics identical to HCP's `NodePool`.

Documentation for these configuration options can be found in HyperShift's API reference: 
- https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.NodePoolManagement exposes `MaxSurge`.
- https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.NodePoolSpec exposes `NodePoolTimeout`

Example `MachineConfigPool` including both `NodeDrainTimeout` and a `MaxSurge` setting:
```yaml
kind: MachineConfigPool
spec:
  # Existing spec fields are not shown.    

  # Adopted from NodePool to create consistency and further our goal
  # to improve the reliability of worker updates.
  nodeDrainTimeout: 10m
      
  # New policy analog to NodePool.NodePoolManagement.
  machineManagement:
    upgradeType: "Replace"
    replace:
      strategy: "RollingUpdate"
      rollingUpdate:
        maxUnavailable: 0
        maxSurge: 4
```

Like HCP, `UpgradeType` will support:

> - `InPlace` where no additional nodes are brought online to support draining workloads.

Reviewer 2: "InPlace" is the way nodes are updated today.  Is that changing for non-HCP?  If yes, where is the reference?  I do not see this in the spec for MCP.  MUO implements surge for the current in-place node updates.  These need to be decoupled, upgrade type independent of max surge configuration.

Reviewer 3: I agree here, I don't think we should switch to having multiple update types as part of this enhancement.

- `Replace` where new nodes will be brought online with `MaxSurge` support.

#### InPlace Update Type

The `InPlace` update type is similar to `MachineConfigPools` traditional behavior where a 
user can configure the `MaxUnavailable` nodes. This approach assumes the number (or percentage)

> of nodes specified by `MaxUnavailable` can be drained simultaneously with workloads finding

Reviewer: What happens here today? Does it do nodes 1 by 1 or in groups? Is there any existing configuration for this?

Reviewer 2: MCP has maxUnavailable today, the difference here would be the scope.  It has been ages since I poked at the behavior of maxUnavailable > 1 for MCP but what I recall is it wouldn't care what zones or machinesets a machine was a member of.  It simply picked machines and went to town against all machines in the pool.  That pool being "workers" meant it could be very disruptive to workloads.

sufficient resources on other nodes to avoid stalled drains.

```yaml
kind: MachineConfigPool
spec:
  # Existing spec fields are not shown.    

  machineManagement:
    upgradeType: "InPlace"
    inPlace:
      maxUnavailable: 10%
```

Nodes are rebooted after they are drained.

#### Replace Update Type

The `Replace` update type removes old machines and replaces them with new instances. It supports
two strategies:

> - `OnDelete` where a new machine is brought online only after the old machine is deleted.

Reviewer: You may want to clarify here that "deleted" means marked for deletion, doesn't mean that the node is gone away. Eg the cluster might not have space to move the evicted pods so a new node is needed to come up before this deleted one can go away

Author: According to https://github.com/kubernetes-sigs/cluster-api/pull/4346 , OnDelete in CAPI implies:
> The machineDeployment controller waits for the machine to get deleted completely and then will proceed to provision the new replica with the new configuration.

This can absolutely lead to wedging in a cluster with insufficient capacity. The main use case I can see here is if you are truly limited to N nodes by your provider and attempts at N+1 would fail indefinitely. `OnDelete` with a drain timeout achieves an update (with disruption) in this scenario, but you can at least get through the update.

Reviewer: Ack, that is not how I remember that working... Hmm, wonder if that changed 🤔 

Then yes, fair with the drain timeout.

Out of interest, what would you expect as an acceptable range of values for the Drain timeout? What if I tried to set it to, 1 year?

Author: > Out of interest, what would you expect as an acceptable range of values for the Drain timeout? What if I tried to set it to, 1 year?

Between 1 second (admin just wants the update to unfold immediately) and int64 seconds (grace period  data type podspec). If podspec doesn't have a perspective on what is reasonable, I don't think we should.

- `RollingUpdate` which supports the `MaxSurge` option.

```yaml
kind: MachineConfigPool
spec:
  # Existing spec fields are not shown.    

  machineManagement:
    upgradeType: "Replace"
    replace:
      strategy: "RollingUpdate"
      rollingUpdate:
        maxUnavailable: 0
        maxSurge: 4
```

`MaxSurge` applies independently to each associated MachineSet. For example, if three MachineSets are associated
with a MachineConfigPool, and `MaxSurge` is set to 4, then it is possible for the cluster to surge up to 12 nodes
(4 for each of the 3 MachineSets).

### Topology Considerations

Multi-AZ (availability zone) clusters function by using one or more `MachineSets` per zone. In order for this enhancement
to work across all zones, all such `MachineSets` should be associated with a `MachineConfigPool` with well considered
values for `MaxSurge` and `NodeDrainTimeout`.

> Each `MachineSet` associated with a `MachineConfigPool` will be permitted to scale by the `MaxSurge` number of nodes. 

Reviewer: This will be confusing for end users, we may want to try and name the field so that it indicates this better than what it does today.

Also, this really makes me think we want to reverse this and look at this as a CAPI problem rather than an MCP problem

There may be pods running on 1b nodes that can only be scheduled on 1b nodes (e.g. due to taints / affinity / 
topology constraints, machine type, etc.). If `MaxSurge` was interpreted in such a way as to only surge machine-set-1a by 1 node,
constrained pods requiring 1b nodes could not benefit from this additional capacity.

Instead, this enhancement proposes each `MachineSet` be permitted to surge up to the `MachineConfigPool` surge
value independently. 

#### Hypershift / Hosted Control Planes

N/A. Hosted Control Planes provide the model for the settings this enhancement seeks to expose in standalone clusters.

#### Standalone Clusters

The `MachineConfigPool` custom resource must be updated to expose the new semantics.  The existing `spec.maxUnavailable`
will be deprecated in favor of the more expressive `MachineManagement` stanza. 

#### Single-node Deployments or MicroShift

N/A.

### Implementation Details/Notes/Constraints

`MachineSets` or CAPI equivalents must support `MachineConfigPool` values for `MaxSurge`. When `MaxSurge` is greater
than 0, the controller can instantiate more machines than the number of specified `MachineSet` replicas. 

> This must work seamlessly with all cluster autoscaler options. The cluster autoscaler may need to be aware of 

Reviewer 2: MUO side-steps this by creating new machinesets for the surge capacity.

Author: Resolving since I've added this as the proposed implementation.

it will not impact drains triggered by the CLI).

### User Stories

Implementing surge and node drain timeout support in `MachineConfigPools` can simplify cluster management for self-managed 
standalone clusters as well as managed clusters (i.e. Service Delivery can remove this customized behavior from the MUO and use
more of the core platform).

* As an Operations team managing one or more standalone clusters, I want to
  help ensure smooth updates by surging my worker node count without constantly 
  having my cluster over-provisioned.
* As an Operations team managing one or more standalone clusters, I want to 
  ensure my cluster update makes steady progress by limiting the amount of time a node drain can
  consume.
* As an engineer in the Service Delivery organization, I want to use core platform
  features instead of developing, evolving, and testing the Managed Upgrade Operator.
* As an Operations team managing standalone and HCP based OpenShift clusters, I
  want a consistent update experience leveraging a surge strategy and/or 
  node drain timeouts regardless of the cluster profile.

### Goals

> - Implement an update configuration, including `MaxSurge`, similar to HCP's `NodePool`

Reviewer 3: This kinda ties into our managed bootimage path https://github.com/openshift/enhancements/blob/e97e6773669d24f2a8dba57ba7c533117ce5f5b3/enhancements/machine-config/manage-boot-images.md, just wanted to cross link those

  in standalone OpenShift's `MachineConfigPool`. 

> - Implement `NodeDrainTimeout`, similar to HCP's `NodePool`, in standalone OpenShift's `MachineConfigPool`.

Reviewer 3: This should be a bit more straightforward to do since it shouldn't be platform specific

  # to improve the reliability of worker updates. This only applies
  # to drains triggered by the MCO (e.g. CLI triggered drains will
  # not be impacted).
  nodeDrainTimeout: 10m
      
  # New policy analog to NodePool.NodePoolManagement.
  machineManagement:
    upgradeType: "Replace"
    replace:
      strategy: "RollingUpdate"
      rollingUpdate:
        maxUnavailable: 0
        maxSurge: 4
```

Like HCP, `UpgradeType` will support:
- `InPlace` where no additional nodes are brought online to support draining workloads.
- `Replace` where new nodes will be brought online with `MaxSurge` support.

#### InPlace Update Type

The `InPlace` update type is similar to `MachineConfigPools` traditional behavior where a 
user can configure the `MaxUnavailable` nodes. This approach assumes the number (or percentage)
of nodes specified by `MaxUnavailable` can be drained simultaneously with workloads finding
sufficient resources on other nodes to avoid stalled drains.

```yaml
kind: MachineConfigPool
spec:
  # Existing spec fields are not shown.    

  machineManagement:
    upgradeType: "InPlace"
    inPlace:
      maxUnavailable: 10%
```

Nodes are rebooted after they are drained.

> #### Replace Update Type

Reviewer 3: There's an implicit difference between HCP and MCO in that HCP's replace will always use the latest payload bootimage, and MCO will always use the install-time bootimage. If we do want to get into the business of update types, maybe we should make bootimage management fully explicit as a dependency of this (see linked enhancement above)
